File: golang/199/199.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func rightSideView(root *TreeNode) []int {
	if root == nil {
		return nil
	}

	q := []*TreeNode{root}

	res := make([]int, 0)
	for len(q) > 0 {
		res = append(res, q[len(q)-1].Val)

		tmp := q
		q = make([]*TreeNode, 0)
		for _, v := range tmp {
			if v.Left != nil {
				q = append(q, v.Left)
			}

			if v.Right != nil {
				q = append(q, v.Right)
			}
		}
	}

	return res
}

/* 方法二

var res [][]int

func rightSideView(root *TreeNode) []int {
	if root == nil {
		return nil
	}

	res = make([][]int, 0)
	levelOrder(root, 0)

	r := make([]int, 0)
	for _, v := range res {
		r = append(r, v[len(v)-1])
	}
	return r
}

func levelOrder(root *TreeNode, level int) {
	if root == nil {
		return
	}

	if level == len(res) {
		res = append(res, []int{})
	}
	res[level] = append(res[level], root.Val)

	levelOrder(root.Left, level+1)
	levelOrder(root.Right, level+1)
}
*/

File: cpp/199.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
private:
    vector< vector<int> > res;
public:
    vector<int> rightSideView(TreeNode* root) {
        levelOrder(root,1);        
        vector<int> vec;
        for (int i=0; i<res.size(); i++){
            vec.push_back(res[i].back());
        }
        return vec;
    }
    
    void levelOrder(TreeNode *root, int level){
        if(!root) return;
        if(res.size()<level){
            res.push_back(vector<int>());
        }
        res[level-1].push_back(root->val);
        levelOrder(root->left, level+1);
        levelOrder(root->right, level+1);
    }
};